A gift to javascript naysayers

juuz0 opened the pull request:
Fixes #615

juuz0 commented:
@kelson42 I haven't written tests yet (still reading through how to write them, should be done soon) but you can test the main functionality. Testing endpoint is `TLD/nojs`

kelson42 commented:
@juuz0 Thank you very much for your PR. I have put @rgaudin as functional reviewer because he has requested the feature and has probably stronger opninion abuot how this shoudl be fixed. I have as well put @veloman-yunkan for the technical review (and functional as well :)

rgaudin commented:
OK, I've tested it and it works fine ; congrats. Couple of comments though:

- how is this going to be used? Right now, one has to go to `/nojs` manually to get this rendition. How will this be integrated for effective use? In a separate endpoint linked in a `<noscript />` on the main page?
- I understand the style was copied from the regular version but the magnification effect on a card's hovering only makes sense because it's clickable (and even then…) but in the nojs version it's not clickable and it thus just a distracting confusion.
- where should those link point to? Currently it points to the viewer… which is JS dependent

@kelson42 what's the merge strategy for this?

veloman-yunkan reviewed (commented):
I skimmed through the changes very quickly and I think that it is yet premature commenting on the code. At this point I will provide high level input:

1. Shouldn't the output be paginated?
2. What about translation? Once #846 is merged you can reuse the same strings for translating in the backend.

mgautierfr commented on src/server/internalServer.cpp:
```diff
@@ -744,6 +747,41 @@ std::unique_ptr<Response> InternalServer::handle_viewer_settings(const RequestCo
+
+  const auto url = request.get_url();
+  const auto domains = kiwix::split(url, "/", true, false);
+  auto filter = get_search_filter(request);
+  try {
+    if (request.get_argument("category") == "") {
+      filter.clearCategory();
+    }
+    if (request.get_argument("lang") == "") {
+      filter.clearLang();
+    }
+  } catch (...) {}
```
I'm surprised about that. In which context `get_search_filter` will set a filter on the category and we need to clear it ?
Shouldn't be better to fix `get_search_filter` if it is buggy ?

juuz0 replied:
My reasoning for not touching `get_search_filter` was if the behaviour was intentional to display no books when lang/category == "" and if I change that logic, I would break some other functionality.
In kiwix-serve welcome page filters, "" means all categories while `get_search_filter` seems to consider "" as a valid category and rightly shows no book associated

juuz0 replied:
I had assumed `category==""` was for all categories filter
Is there a specific category for "all categories" then?

mgautierfr replied:
I'm not sure about what we are speaking about.
But if we are speaking about the request send to the http api, if we don't want to filter on the category, we must not pass a `category==""` query string.

juuz0 replied:
There needs to be a way for the server to know that the user requests "ALL CATEGORIES", I dont think a query param can be cleared without javascript on the client side hence I consider `category=""` as ALL categories & clear that on the server side.

mgautierfr replied:
> There needs to be a way for the server to know that the user requests "ALL CATEGORIES"

The absence of filter on category seems to me a good way to specify all categories.

> I dont think a query param can be cleared without javascript on the client side hence I consider category="" as ALL categories & clear that on the server side.

Why we need to clear it ?
The html code is generated server side, so we can generate the request we want, without a `category` parameter.

veloman-yunkan replied:
> > I dont think a query param can be cleared without javascript on the client side hence I consider category="" as ALL categories & clear that on the server side.
> 
> Why we need to clear it ? The html code is generated server side, so we can generate the request we want, without a `category` parameter.

On the server side we generate the HTML for the search form. The form is filled and submitted from the client side. Values are sent for all inputs of the form (unless it is a request generated via JS).

mgautierfr replied:
Ha yes, I've forget about the search form.
I would say that `get_search_filter` should be the only (and correct) way to create a filter from query string.
If we add a modification on the filter in specific api, it will introduce inconsistency in the API and we can be sure that we will mix things later. 

> while get_search_filter seems to consider "" as a valid category and rightly shows no book associated.

I finally think it is wrong. The only valid use case of searching for no category is to search for book which have no category associated. So it is more something about `undefined` category than <code></code> category.

Maybe we could extend a bit the api to accept to specific values `[all]` and `[empty]` (or `[none]`) for no filtering (all) and filtering for book without category (empty/none).

The `[]` would prevent a potential conflict with a real category `all` or `none`. 
What do you think about that ?

juuz0 replied:
Extending the API would make life easier definitely & bring more consistency.
Would've been nice if `all`/`none` be reserved words for filters but I don't really mind the `[all]` or `[empty/none]` (`none` is better word imo).
Only consideration would be if this looks pleasant on the eye in URL: `TLD/?lang=[all]&category=[none]`

mgautierfr commented on src/server/internalServer.cpp:
```diff
@@ -744,6 +747,41 @@ std::unique_ptr<Response> InternalServer::handle_viewer_settings(const RequestCo
+  OPDSDumper opdsDumper(mp_library, mp_nameMapper);
+  opdsDumper.setRootLocation(m_root);
+  opdsDumper.setLibraryId(getLibraryId());
+  std::string content = opdsDumper.noJSData(filter);
+
+  if (domains.size() == 1) {
+    content = opdsDumper.noJSData(filter);
+  } else if ((domains.size() == 3) && (domains[1] == "download") && (mp_nameMapper->getIdForName(domains[2]) != "")) {
+    content = opdsDumper.noJSDownloadLinks(mp_nameMapper->getIdForName(domains[2]));
+  } else {
+    return HTTP404Response(*this, request)
+           + urlNotFoundMsg;
+  }
```
This set the content to the result of `opdsDumper.noJSData(filter)` all the time and then overwrite it.
So the first initialization is useless. And we should create the filter only if `domain.size() == 1`, not all the time.

juuz0 commented:
> how is this going to be used? Right now, one has to go to /nojs manually to get this rendition. How will this be integrated for effective use? In a separate endpoint linked in a <noscript /> on the main page?

Hiding nav, body, footer when javascript is not enabled & displaying a text "This page can only be accessed with Javascript enabled, please head over to `<a href="{{root}}/nojs">/nojs endpoint</a>` should be fine I think?

> I understand the style was copied from the regular version but the magnification effect on a card's hovering only makes sense because it's clickable (and even then…) but in the nojs version it's not clickable and it thus just a distracting confusion.

It is clickable though? I had to reduce clickable area (currently it's only the description part) because HTML spec doesn't allow `<a>` nesting (another `<a>` is used for download button)

> where should those link point to? Currently it points to the viewer… which is JS dependent

Will go by mgautierfr's suggestion, thanks! 

> Shouldn't the output be paginated?

normal welcome page (JS one) isn't paginated so why this? Performance reasons?

> What about translation? Once Internationalization of kiwix-serve's frontend #846 is merged you can reuse the same strings for translating in the backend.

Noted, should be done soon - once I understand how to write tests which is taking some effort :)

rgaudin commented:
> Hiding nav, body, footer when javascript is not enabled & displaying a text "This page can only be accessed with Javascript enabled, please head over to `<a href="{{root}}/nojs">/nojs endpoint</a>` should be fine I think?

Fine by me. What do others think about?

> It is clickable though? I had to reduce clickable area (currently it's only the description part) because HTML spec doesn't allow `<a>` nesting (another `<a>` is used for download button)

No it's not clickable. That's the point, it makes user think it is because it grows

> Will go by mgautierfr's suggestion, thanks!

Way to go. This endpoint also handles the mainPath so that should be easy… although this URL should normally be retrieved from OPDS I suppose…

> normal welcome page (JS one) isn't paginated so why this? Performance reasons?

It is. Just look at https://library.kiwix.org/ there's an automatic loading of new cards on scroll. I'm not in favor or paginating for a first version. It's unlikely this will be really useful in practice and adds some complexity. Let's wait for user requests for this.

juuz0 commented:
> how is this going to be used?......

Done as discussed.

> I understand the style was copied from the regular....

Removed hover effect

> where should those link point to? Currently it points to the viewer....

Done.

> What about translation? .....

Still pending, I'm so scared to rebase this PR on top of main currently. Should be done soon.

Added two tests

juuz0 commented:
dunno why CI is failing :(

juuz0 commented:
> What about translation? Once 846 is merged you can reuse the same strings for translating in the backend.

Done :)

juuz0 commented:
@rgaudin ping because there's no button to re request a review! >:)

juuz0 commented:
For translations, use `userlang` as url param.
Example
`http://192.168.18.8:8080/nojs?lang=eng&userlang=test`

rgaudin reviewed (changes requested):
Just tested 8c10936
- Home points to `/nojs` as expected 👍 
- Links points to `/content` 👍 
- i18n translation works fine 👍
- I still get the hover effect on the Tag: text is larger, bold and the pointer changes as well. ❌

juuz0 commented:
> I still get the hover effect on the Tag: text is larger, bold and the pointer changes as well. x

Done now, thanks for noticing.

Additionally, I had forgotten to display "No results" when there were 0 books. That's done too. Along with a new test

rgaudin commented:
> > I still get the hover effect on the Tag: text is larger, bold and the pointer changes as well. x
> 
> Done now, thanks for noticing.

👍 OK now

veloman-yunkan commented on src/opds_dumper.cpp:
```diff
@@ -282,12 +282,14 @@ std::string OPDSDumper::noJSData() const
   for ( const auto& bookId : library->getBooksIds() ) {
     const auto bookObj = library->getBookById(bookId);
     const auto bookTitle = bookObj.getTitle();
+    const auto contentId = bookObj.getHumanReadableIdFromPath();
```
Obtain the user-friendly content id via `nameMapper`

veloman-yunkan commented on static/templates/no_js_download.html:
```diff
@@ -0,0 +1,23 @@
+    <meta name="viewport" content="width=device-width, initial-scale=1.0">
+    <title>Download links</title>
+</head>
+<body>
```
Shouldn't this page contain the name of the book?

veloman-yunkan commented on src/opds_dumper.cpp:
```diff
@@ -276,6 +276,17 @@ kainjow::mustache::list getTagList(std::string tags)
+               {"url", book.getUrl() }
+             }
+  );
+}
```
Maybe this method should belong to `InternalServer`?

veloman-yunkan commented on src/opds_dumper.cpp:
```diff
@@ -287,10 +287,13 @@ std::string OPDSDumper::noJSDownloadLinks(const std::string& bookId) const
 }
 
-std::string OPDSDumper::noJSData() const
+std::string OPDSDumper::noJSData(kiwix::Filter filter) const
```
The API of dumping in plain HTML format diverges from the API of dumping in OPDS format. Why not stick to the signature of the `dumpOPDSFeed2()` and obtain the book ids in `InternalServer` (probably using `InternalServer::search_catalog()`)?

juuz0 replied:
`InternalServer::search_catalog()` doesnt allow me to change the filter :<< I need to clear the category

veloman-yunkan commented on static/templates/no_js.html:
```diff
@@ -4,7 +4,7 @@
+    <title>{{welcome-to-kiwix-server}}</title>
     <link
       type="text/css"
       href="{{root}}/skin/index.css?KIWIXCACHEID"
```
Passing translation strings as top-level mustache variables feels somewhat fragile to me. Can't all translations required for this template be passed as one `translations` object (so that they are accessed as `{{translations.welcome-to-kiwix-server}}`? I am not sure that it will work but I believe that it should.

juuz0 replied:
Done. Thanks

veloman-yunkan commented on test/library_server.cpp:
```diff
@@ -941,6 +941,221 @@ TEST_F(LibraryServerTest, no_name_mapper_catalog_v2_individual_entry_access)
+  "            </div>\n" \
+  "        </div>\n" \
+  "        </div>\n"
+#define RAY_CHARLES_UNCTZ \
```
Add `_BOOK_HTML` for consistency

veloman-yunkan reviewed (changes requested):
This is a first pass. There are some comments that imply a discussion and a final decision together with @mgautierfr before the proposed changes are made.

mgautierfr commented on src/opds_dumper.cpp:
```diff
@@ -262,6 +262,25 @@ std::string OPDSDumper::categoriesOPDSFeed() const
   );
 }
 
+std::string OPDSDumper::noJSData() const
```
This is a interesting point. We could indeed have a `LibraryDumper` (may be only a base class if implementation need it) and have the opds, plain html and libxml dumping using it.
All three dumping are a dumping of the library in different formats. If we can do it only with templates it would be nice, if we cannot ... we cannot. But it would be nice to have the common parts (to be defined) put in common code.

At least, the renaming proposed by @veloman-yunkan seems good.

mgautierfr replied:
This seems good to me.

(comment based only on the new organisation of  `LibraryDumper`, `OPDSDumper` & `HTMLDumper`. Some other comments may still apply on the code itself)

veloman-yunkan replied:
It seems good to me, too.

Can you please rebase and update the history of this PR by extracting `LibraryDumper` out of `OPDSDumper` as the first step and then adding `HTMLDumper`?

juuz0 replied:
Done :>

juuz0 commented on test/library_server.cpp:
```diff
@@ -941,6 +941,221 @@ TEST_F(LibraryServerTest, no_name_mapper_catalog_v2_individual_entry_access)
+  "    <style>\n" \
+  "      @font-face {\n" \
+  "        font-family: \"poppins\";\n" \
+  "        src: url(\"/skin/fonts/Poppins.ttf?cacheid=af705837\") format(\"truetype\");\n" \
```
Fixed the bug

veloman-yunkan commented on src/opds_dumper.cpp:
```diff
@@ -341,34 +341,33 @@ std::string OPDSDumper::noJSDownloadLinks(const std::string& bookId) const
 }
 
-std::string OPDSDumper::noJSData(kiwix::Filter filter) const
+std::string OPDSDumper::dumpPlainHTML(const std::vector<std::string>& bookIds, const std::string& query,
+                                const std::string& category, const std::string& language) const
```
With `HTMLDumper` now being separate from `OPDSDumper` the difference in the signatures of the dump functions doesn't matter that much. The version taking a single `Filter` parameter was better.

juuz0 replied:
Done.

veloman-yunkan reviewed (changes requested):
I would like to review a clean version of this PR that starts with `LibraryDumper` being extracted out of `OPDSDumper` before `HTMLDumper` is introduced in the second commit.

juuz0 commented:
I haven't added the long copyright text in `library_dumper.*` or `html_dumper.*`, should i just copy paste that text from `opds_dumper`?

juuz0 commented on include/library_dumper.h:
```diff
@@ -0,0 +1,91 @@
+class LibraryDumper
+{
+ public:
+  LibraryDumper() = default;
```
the default constructor for OPDSDumper comes from https://github.com/kiwix/libkiwix/commit/b48428e4436a66dd28a9c7279eb3e6efbc681348 
Though, the setLibrary function has been removed now and the default constructor is not used anywhere now. Safe to remove I guess?

veloman-yunkan replied:
Other places to look for are projects depending on libkiwix. I think we should remove the default constructor of `OPDSDumper` in any case, but at least we must know in advance how many things will break and will need to be fixed.

juuz0 commented on src/library_dumper.cpp:
```diff
@@ -0,0 +1,126 @@
+  {"hbs", "srpskohrvatski"},
+  {"ido", "ido"},
+  {"kbp", "kabɩyɛ"},
+  {"kld", "Gamilaraay"},
+  {"lbe", "лакку маз"},
+  {"lbj", "ལ་དྭགས་སྐད་"},
+  {"map", "Austronesian"},
+  {"mhr", "марий йылме"},
+  {"mnw", "ဘာသာမန်"},
+  {"myn", "mayan"},
+  {"nah", "nahuatl"},
+  {"nai", "north American Indian"},
+  {"nds", "plattdütsch"},
+  {"nrm", "bhasa narom"},
+  {"olo", "livvi"},
+  {"pih", "Pitcairn-Norfolk"},
+  {"pnb", "Western Panjabi"},
+  {"rmr", "Caló"},
+  {"rmy", "romani shib"},
+  {"roa", "romance languages"},
+  {"twi", "twi"},
+};
+
+std::once_flag fillLanguagesFlag;
+
+void fillLanguagesMap()
+{
+  for (auto icuLangPtr = icu::Locale::getISOLanguages(); *icuLangPtr != NULL; ++icuLangPtr) {
+    const ICULanguageInfo lang(*icuLangPtr);
+    iso639_3.insert({lang.iso3Code(), lang.selfName()});
+  }
+}
+
+std::string getLanguageSelfName(const std::string& lang) {
+  const auto itr = iso639_3.find(lang);
+  if (itr != iso639_3.end()) {
+    return itr->second;
+  }
+  return lang;
+};
```
Out of curiosity, what does this help in?

veloman-yunkan replied:
It prevents name conflicts (symbol redefinition errors) at link time - this is especially important for libraries since they should ideally only export symbols representing their public API.

juuz0 commented on src/server/internalServer.cpp:
```diff
@@ -53,6 +53,7 @@ extern "C" {
 #include "name_mapper.h"
 #include "search_renderer.h"
 #include "opds_dumper.h"
+#include "library_dumper.h"
```
Not needed, got mixed while i was changing commit history :>

juuz0 replied:
This change doesn't have a fixup associated, but fixed by editing commit

juuz0 commented on include/html_dumper.h:
```diff
@@ -23,6 +23,13 @@ class HTMLDumper : public LibraryDumper
+   *
+   * @return HTML content
+   */
+  std::string dumpPlainHTML() const;
```
Does this mean to combine the 2 commits (introducing HTMLDumper and /nojs)?
dumpPlainHTML is sort of dependent on having no_js file

veloman-yunkan replied:
> Does this mean to combine the 2 commits (introducing HTMLDumper and /nojs)?

No. Only include in the commit introducing `HTMLDumper` what is needed by (the initial version of) its `dumpPlainHTML()` method. The name of the new template `no_js.html` doesn't make it an unseparable part of the change that introduces the `/nojs` endpoint (just suppose that the template is named `nojs_library_page.html`; BTW, isn't that a better filename?)

juuz0 replied:
This change doesn't have a fixup associated, but fixed by editing commit

juuz0 commented on static/skin/i18n/en.json:
```diff
@@ -28,7 +28,7 @@
 	, "searchbox-tooltip": "Search '{{BOOK_TITLE}}'"
 	, "confusion-of-tongues": "Two or more books in different languages would participate in search, which may lead to confusing results."
-	, "welcome-page-overzealous-filter": "No result. Would you like to <a href=\"?lang=\">reset filter</a>?"
+	, "welcome-page-overzealous-filter": "No result. Would you like to <a href=\"?lang=&category=&q=\">reset filter</a>?"
```
It already handles the absence, I thought it was more consistent to give all params that time - I don't think so anymore. Will reset it.
Atlhough, with #907 changing the `?` to `#`, I think we need a overzealous-filter string for `?`, correct?

veloman-yunkan replied:
> with #907 changing the `?` to `#`, I think we need a overzealous-filter string for `?`, correct?

Sorry, but I don't understand your question/proposal.

juuz0 replied:
https://github.com/kiwix/libkiwix/pull/907/files#diff-c70f0ae73979b109350424912c766a72199a66d69ecfa14c1afbe54dc9672ceeR31

Once this change is merged, we'll be using fragment params in js-enabled welcome page while query params in no-js page. My question was if I should be adding a new `welcome-page-overzealous-filter-query` entry with `<a href=\"?lang=\">` as link instead of `<a href=\"#lang=\">`

juuz0 replied:
Very nice :D Thanks
Added in https://github.com/kiwix/libkiwix/pull/897/commits/eac892e52330b107948082ec31f3da8babdf05d8

P.S. #907 should be merged before this PR

veloman-yunkan commented on include/library_dumper.h:
```diff
@@ -77,8 +94,8 @@ class LibraryDumper
-  kiwix::Library* library;
-  kiwix::NameMapper* nameMapper;
+  kiwix::Library* const library;
+  kiwix::NameMapper* const nameMapper;
```
`const` was added in the wrong place. I intended `const Library* library;` and `const NameMapper* nameMapper;` (we don't want `LibraryDumper` to be able to call non-const operations on library and/or name mapper objects). Though the immutability of those data-members doesn't hurt either. So let's end up with

```
const kiwix::Library* const library;
const kiwix::NameMapper* const nameMapper;
```

juuz0 replied:
Apologies :)
I always mess up between `pointer to const` and `const pointer` :P

veloman-yunkan commented on src/server/internalServer.cpp:
```diff
@@ -759,10 +759,22 @@ std::unique_ptr<Response> InternalServer::handle_viewer_settings(const RequestCo
   return ContentResponse::build(*this, RESOURCE::templates::viewer_settings_js, data, "application/javascript; charset=utf-8");
 }
 
+std::string InternalServer::dumpDownloadLinks(const std::string& bookId) const
```
In the content of `InternalServer` the verb `dump` sounds a little strange. `getNoJSDownloadPageHTML()` is a better name.

veloman-yunkan commented on src/server/i18n.h:
```diff
@@ -65,6 +65,17 @@ class GetTranslatedString
     return expandParameterizedString(m_lang, key, params);
   }
 
+  std::pair<std::string, kainjow::mustache::basic_data<std::string>> getTranslatedPair(std::string key)
+  {
+    return {key, getTranslatedString(m_lang, key)};
+  }
+
+  std::pair<std::string, kainjow::mustache::basic_data<std::string>> getTranslatedPair(std::string key, kainjow::mustache::object params)
+  {
+    return {key, expandParameterizedString(m_lang, key, params)};
+  }
+
+
```
1. Make the signatures consistent with those of preexisting functions.

2. It's better to introduce a separate class `GetTranslatedStringWithMsgId` with a pair of overloaded `operator()`s instead of those functions.

veloman-yunkan commented on src/server/internalServer.cpp:
```diff
@@ -759,14 +759,25 @@ std::unique_ptr<Response> InternalServer::handle_viewer_settings(const RequestCo
   return ContentResponse::build(*this, RESOURCE::templates::viewer_settings_js, data, "application/javascript; charset=utf-8");
 }
 
+std::string userLang = "eng";
```
A mutable global variable is an issue for concurrent access.

juuz0 replied:
Made it a member of internalServer, i've a feeling there are better ways but if this works then nice!

veloman-yunkan replied:
The more general problem for concurrent access is a shared mutable variable. A global variable merely has more chances of becoming one. Your "fix" doesn't prevent m_userLang from being changed from two or more different threads during concurrent execution of multiple `/nosj` requests. However, it is not clear why you needed to make `userLang` a data member of `InternalServer` at all. A local variable should be just fine.

Since I haven't yet looked at your other changes, please correct this in place (i.e. without introducing a fixup commit).

juuz0 replied:
Without having some sort of "shared" access to m_userLang between HTMLDumper & InternalServer - the `/nojs/download` page wouldn't know what userlang to display. Hence the requirement of a member `m_userLang`.

Consider this:

`/nojs?userlang=test`
Click a book's download button & you're redirected to `/nojs/download/xyz` - if we're using a local variable, it would ask for `userlang` argument again - not get it & set the language to default English.


Another solution is probably setting all download links to have userlang as query. Wouldn't require any member in internalServer then, is this better?

veloman-yunkan replied:
You must not store any state *in the server* between requests (in this case, with proper timing `m_userLang` set by one user's access to `/nojs` may affect the language used for serving `/nojs/download/xyz` to another user). 

Note that the `userlang` URL query parameter is only one way of specifying the user language. The user-language (no matter where it comes from) is stored in a cookie:
https://github.com/kiwix/libkiwix/blob/88de978a9c52aa8e64dad883e949be1e9250dfa7/src/server/request_context.cpp#L210-L228
https://github.com/kiwix/libkiwix/blob/88de978a9c52aa8e64dad883e949be1e9250dfa7/src/server/response.cpp#L390-L395

juuz0 replied:
Done

veloman-yunkan reviewed (changes requested):
Please include testing of the `/nojs` endpoint in the `ServerTest.CacheControlOfDynamicContent` unit-test.

Also, please rebase&fix-up for the next iteration.

veloman-yunkan commented on src/html_dumper.cpp:
```diff
@@ -74,15 +75,27 @@ std::string HTMLDumper::dumpPlainHTML(kiwix::Filter filter) const
     });
   }
 
+  auto i18nTranslations = i18n::GetTranslatedStringWithMsgId(m_userLang);
```
`i18nTranslations` -> `getTranslation`

veloman-yunkan commented on src/server/internalServer.cpp:
```diff
@@ -759,14 +759,23 @@ std::unique_ptr<Response> InternalServer::handle_viewer_settings(const RequestCo
+std::string InternalServer::getNoJSDownloadPageHTML(const std::string& bookId, const std::string& userLang) const
 {
   const auto book = mp_library->getBookById(bookId);
+  auto i18nTranslations = i18n::GetTranslatedStringWithMsgId(userLang);
```
`i18nTranslations` -> `getTranslation`

veloman-yunkan reviewed (changes requested):
Good. There are a few things to fix to keep the code tidy but else we are almost done.

juuz0 commented on include/opds_dumper.h:
```diff
@@ -38,10 +39,9 @@ namespace kiwix
 {
  public:
-  OPDSDumper() = default;
   OPDSDumper(Library* library, NameMapper* NameMapper);
```
I haven't heard that term before, does this mean using `const_cast` to remove constness?

veloman-yunkan replied:
`OPDSDumper` is now more demanding than it has to be - if one has a `const Library` object and/or a `const NameMapper` object they cannot create an `OPDSDumper` (without resorting to a `const_cast`, which is bad manners). There is no conceptual or technical reason why the current signature of the `OPDSDumper` constructor (unlike `HTMLDumper`) prohibits such an invocation. That's why I requested to *relax* the too strict requirements that exits merely for historical reasons.

veloman-yunkan commented on src/html_dumper.cpp:
```diff
@@ -35,12 +35,17 @@ std::string HTMLDumper::dumpPlainHTML() const
     const auto tags = bookObj.getTags();
     std::string faviconAttr = "style=background-image:url(" + bookIconUrl + ")";
     booksData.push_back(kainjow::mustache::object{
+      {"id", contentId},
```
`contentId` must be URI-encoded

juuz0 commented on static/templates/no_js_library_page.html:
```diff
@@ -0,0 +1,140 @@
+      }
+
+      @font-face {
+          font-family: "roboto";
+          src: url("{{root}}/skin/fonts/Roboto.ttf?KIWIXCACHEID") format("truetype");
+      }
+
+      .book__list {
+        display: flex;
+        flex-direction: row;
+        flex-wrap: wrap;
+        align-items: center;
+      }
+
+      .book__wrapper:hover {
+        transform: scale(1.0);
+      }
+
+      .tag__link {
+        pointer-events: none;
+      }
+
+      .book__link__wrapper {
+        grid-column: 1 / 3;
+        grid-row: 1 / 3;
+      }
+
+      .book__link {
+        grid-row: 2 / 3;
+      }
+
+      .kiwixHomeBody__results {
+        flex-basis: 100%;
+      }
+
+      #book__title>a, .book__download a {
+        text-decoration: none;
+        all: unset;
+      }
+    </style>
```
`index.css` is reused in no_js_library_page.html.
These changes in <style> are specific to no-js

juuz0 commented on include/library_dumper.h:
```diff
@@ -0,0 +1,102 @@
+   * 
+   * @param title 
+   */
+  std::string humanFriendlyTitle(std::string title) const;
```
> Also does this helper function need to be a member of LibraryDumper? Can't it be a free function in an unnamed namespace in html_dumper.cpp?

I thought it would come to use as  a shared function but eh I guess not now, changed as per suggestion.

juuz0 commented on static/skin/i18n/en.json:
```diff
@@ -46,4 +46,6 @@
 	, "filter-by-tag": "Filter by tag \"{{TAG}}\""
 	, "stop-filtering-by-tag": "Stop filtering by tag \"{{TAG}}\""
+	, "welcome-to-kiwix-server": "Welcome to Kiwix Server"
+	, "download-links-title": "Download links for {{BOOK_TITLE}}"
```
This title goes into the tab, dont think styling works there/is needed
![image](https://user-images.githubusercontent.com/26814569/226177736-276b04c0-16a8-4a5f-b7e1-6ad9ca5b77ae.png)

veloman-yunkan replied:
Well, this is the first time I ran the new code. I was under the wrong impression that this is shown as the heading of the page which I believe is the right thing to do. The title of the tab is not a good place (as the only place) for such information.

![Screenshot from 2023-03-20 17-40-44](https://user-images.githubusercontent.com/11933502/226357384-65e9d07b-df9d-4fa7-a8ac-204705199ea3.png)

juuz0 replied:
Done.

juuz0 commented on test/server.cpp:
```diff
@@ -1522,6 +1718,38 @@ TEST_F(ServerTest, CacheControlOfDynamicContent)
+            FILTERS_HTML("")
+            HOME_BODY_0_RESULTS
+            FINAL_HTML_TEXT);
 }
```
I created a new file `test/html_dumper.cpp` for nojs tests (though, not too sure if that name is ok)

> In https://github.com/kiwix/libkiwix/pull/897#pullrequestreview-1334738349 I meant that the ServerTest.CacheControlOfDynamicContent unit-test must be enhanced to cover this new /nojs endpoint too.

I don't know how to do this, but added /nojs in `resources200Compressible`. If that's what was needed, cool! otherwise some guidance please! :>

veloman-yunkan replied:
> I created a new file `test/html_dumper.cpp` for nojs tests (though, not too sure if that name is ok)

No, that's an overshoot. A new test-case in `test/server.cpp` is fine (but it must be placed reasonably).

> 
> > In [#897 (review)](https://github.com/kiwix/libkiwix/pull/897#pullrequestreview-1334738349) I meant that the ServerTest.CacheControlOfDynamicContent unit-test must be enhanced to cover this new /nojs endpoint too.
> 
> I don't know how to do this, but added /nojs in `resources200Compressible`. If that's what was needed, cool!

Cool! I wonder if ChatGPT would be able to do that (I hope that not).

juuz0 replied:
> No, that's an overshoot. A new test-case in test/server.cpp is fine (but it must be placed reasonably).

Okie removed and added them in a new test :> I hope this back and forth stuff hasn't annoyed you (yet!) :P

> Cool! I wonder if ChatGPT would be able to do that (I hope that not).

Hoping the same :P
But also waiting for the day when I tell GPT4 to do welcome page UI from a [paper drawing like this](https://twitter.com/heykahn/status/1635752848398102530)

juuz0 commented:
for the wrong zim, download page was giving 500 instead of 400 - fixed in https://github.com/kiwix/libkiwix/pull/897/commits/de29f7c7f3a1a72bd9f9c34d1752320e09da736a

> at least in ServerTest.noJS and ServerTest.CacheControlOfDynamicContent, maybe also a negative test in ServerTest.404

Done.

veloman-yunkan commented on test/server.cpp:
```diff
@@ -2008,3 +2014,273 @@ R"(const viewerSettings = {
+  "        <div>Magnet link</div>\n" \
+  "    </a>\n" \
+  "    <a href=\".torrent\" download>\n" \
+  "        <div></div>\n" \
```
The bug made its way into the unit-tests :)

Another thing that bothers me is the values of the `href` attributes in this HTML. Can we make the test more realistic? Probably it can be done easily by moving the noJS test into `test/library_server.cpp`.

juuz0 replied:
Ye done! Thanks

veloman-yunkan reviewed (changes requested):
I hope that this was the last complete review of this PR. Please address the comments using fixup commits.

juuz0 commented on test/server.cpp:
```diff
@@ -284,7 +288,7 @@ R"EXPECTEDRESULT(      href="/ROOT%23%3F/skin/index.css?cacheid=4974c58b"
     <script src="/ROOT%23%3F/skin/isotope.pkgd.min.js?cacheid=2e48d392" defer></script>
     <script src="/ROOT%23%3F/skin/iso6391To3.js?cacheid=ecde2bb3"></script>
-    <script type="text/javascript" src="/ROOT%23%3F/skin/index.js?cacheid=8e06e864" defer></script>
+    <script type="text/javascript" src="/ROOT%23%3F/skin/index.js?cacheid=c0c69750" defer></script>
```
Only this change is done through a rebase

juuz0 commented on static/templates/no_js_library_page.html:
```diff
@@ -51,7 +51,7 @@
         {{#books}}
         <div class="book__wrapper">
-            <a class="book__link" href="#" data-hover="Preview">
+            <a class="book__link" href="{{root}}/content/{{id}}" data-hover="Preview">
```
Replaced this with title & aria-label.

> Is "Preview" a good hint text?

Good enough to me to be honest. `Preview book` or `Open book` may also fit.

For the full JS version, I think its better to do in a new ticket? Or should I do that here too

veloman-yunkan commented:
Please, also update `static/skin/i18n/qqq.json` (i.e. document the new messages there)

juuz0 commented on src/server/internalServer.cpp:
```diff
@@ -759,19 +759,58 @@ std::unique_ptr<Response> InternalServer::handle_viewer_settings(const RequestCo
+    if (subUrl == ".meta4") {
+      bookUrl = bookUrl.substr(0, bookUrl.size() - 6);
+    }
+  }
```
added in stringTools with a new test for it
